core/src/main/java/io/kestra/core/storages/StateStore.java

```diff
@@ -0,0 +1,100 @@
+package io.kestra.core.storages;
+
+import io.kestra.core.exceptions.ResourceExpiredException;
+import io.kestra.core.runners.RunContext;
+import io.kestra.core.storages.kv.KVValue;
+import io.kestra.core.storages.kv.KVValueAndMetadata;
+import io.kestra.core.utils.Hashing;
+import io.kestra.core.utils.Slugify;
+import jakarta.annotation.Nullable;
+
+import java.io.ByteArrayInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.Objects;
+import java.util.Optional;
+
+public record StateStore(RunContext runContext) {
+    /**
+     * Gets the state for the given state name, sub-name and task run value.
+     * @param stateName state name
+     * @param stateSubName state sub-name (optional)
+     * @param taskRunValue task run value
+     * @return an InputStream of the state data
+     */
+    public InputStream getState(String stateName, @Nullable String stateSubName, String taskRunValue) throws IOException, ResourceExpiredException {
+        RunContext.FlowInfo flowInfo = runContext.flowInfo();;
+        try {
+            // We relocate the old state store file to KV Store
+            InputStream file = runContext.storage().getFile(StateStore.oldStateStoreUri(flowInfo.namespace(), flowInfo.id(), stateName, taskRunValue, stateSubName));
+            this.deleteOldStateStoreFile(stateName, stateSubName, taskRunValue);
+            byte[] bytes = file.readAllBytes();
+
+            this.putState(stateName, stateSubName, taskRunValue, bytes);
+            return new ByteArrayInputStream(bytes);
+        } catch (IOException e) {
```

Comment:
Exception should not be used as a control flow, this is not performant and error prone.
Please check that the file exist instead of using a catch.

Reply:
Idk I did that to avoid the overhead of having 2 calls in case it exists as I have to read the file anyway (though I can restrict it to FileNotFoundException) but if you still think it's bad I can change it